Add Duty to Refer and Commissioned Rehabilitation Service screens

bryangaledxw opened the pull request:
# Changes in this PR

* Adds five new pages for entering DTR/CRS info
* This is the first task in Apply to have a branching journey
* This does not address answers from an unused branch possibly appearing in the CYA screen. This will be address in - https://trello.com/c/CwMXhaNT/1191-check-your-answers-handles-branching-tasks

## Screenshots of UI changes

![localhost_3000_applications_ab95ef18-f274-4f92-bbc0-a55fa45f3fa6](https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/assets/94137563/677cd84a-064d-46be-8155-45761002bfa1)
![localhost_3000_applications_ab95ef18-f274-4f92-bbc0-a55fa45f3fa6_tasks_accommodation-referral-details_pages_dtr-submitted](https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/assets/94137563/70a91c10-c243-46c2-b71e-b81bb6978bd9)
![localhost_3000_applications_ab95ef18-f274-4f92-bbc0-a55fa45f3fa6_tasks_accommodation-referral-details_pages_dtr-details](https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/assets/94137563/cd5d7843-597e-44e5-846f-fcde445eec02)
![localhost_3000_applications_ab95ef18-f274-4f92-bbc0-a55fa45f3fa6_tasks_accommodation-referral-details_pages_crs-submitted](https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/assets/94137563/88660738-2d25-413b-aeac-f8a7fb15896b)
![localhost_3000_applications_ab95ef18-f274-4f92-bbc0-a55fa45f3fa6_tasks_accommodation-referral-details_pages_crs-details](https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/assets/94137563/2f10faed-764a-4d33-ab43-c1a1cbc5c5f2)
![localhost_3000_applications_ab95ef18-f274-4f92-bbc0-a55fa45f3fa6_tasks_accommodation-referral-details_pages_other-accommodation-options](https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/assets/94137563/9b1cd6dc-7b8d-4970-96b7-d0304d1a667a)
![localhost_3000_applications_ab95ef18-f274-4f92-bbc0-a55fa45f3fa6_tasks_check-your-answers_pages_review](https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/assets/94137563/984a336a-5d9a-45a2-8dd9-057a439e60be)

# Release checklist

[Release process documentation](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/edit-v2/4247847062?draftShareId=a1c360ab-bd31-4db1-aae3-7cc002761de9).

## Pre-merge checklist

- [ ] Are any changes required to dependent services for this change to work?
  (eg. CAS API)
- [ ] Have they been released to production already?

## Post-merge checklist

- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to test
- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to preprod
- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to prod

<!-- Should a release fail at any step, you as the author should now lead the work to
fix it as soon as possible. You can monitor deployment failures in CircleCI
itself and application errors are found in
[Sentry](https://ministryofjustice.sentry.io/issues/?project=4504129156218880&referrer=sidebar&statsPeriod=24h).
Both events should be automatically sent to our [Slack
channel](https://mojdt.slack.com/archives/C048BJS7S2F). -->

bryangaledxw committed:
Move ContactDetails base class

We will soon have Apply pages that want to use this base class outside of the contact details task, so we move this and the associated Nunjucks partial to a better shared location for use by pages in different tasks

bryangaledxw committed:
Add DTR submitted Apply page

Note that the next() destination is dependent on the user's input on this page

bryangaledxw committed:
Add DTR details Apply page

bryangaledxw committed:
Add CRS submitted Apply page

Note that the next() destination is dependent on the user's input on this page. And the previous() destination is dependent on the user's input on a previous page. We add a helper function, hasSubmittedDtr(), to support this

bryangaledxw committed:
Add CRS details Apply page

bryangaledxw committed:
Add other accommodation options Apply page

Note the previous() destination is dependent on the user's input on a previous page. We add a helper function, hasSubmittedCrs(), to support this

bryangaledxw committed:
Add accommodation referral details pages to Apply journey

We add an accommodation referral details section, containing our five new pages. We can now fully integration test our new pages

bryangaledxw committed:
Update getProbationPractitionerName()

We remove an unused optional parameter. This parameter was originally here to keep this function consistent with getArrivalDate(), but these have very different use cases, so divergence is fine. We'll soon be adding more similar functions alongside getProbationPractitionerName() and we don't want to keep this cruft around for all of those

bryangaledxw committed:
Add completeYesNoInputWithDetailFromPageBody() integration test helper

A common pattern in our Apply form pages is two radio buttons labelled "Yes" and "No”, with a requirement to enter further detail if "Yes" is selected. We have some infrastructure for supporting this case imported from Approved Premises, but nothing for it specifically in our integration tests. We add this helper function for when this pattern arises

bryangaledxw committed:
Correct grammar in unit test names

bryangaledxw committed:
Add dateIsInFuture() helper method

Similar to dateIsInPast(), we add dateIsInFuture() for Apply form validation

bryangaledxw commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Add Duty to Refer and Commissioned Rehabilitation Service screens](https://trello.com/c/Tx9aKiGw/1184-add-duty-to-refer-and-commissioned-rehabilitation-service-screens)

tahb commented on server/form-pages/apply/referrals-and-documents/accommodation-referral-details/crsDetails.ts:
```diff
@@ -0,0 +1,11 @@
+import ContactDetails from '../../contactDetails'
+
+@Page({ name: 'crs-details', bodyProperties: ['name', 'phone', 'email'] })
+export default class CrsDetails extends ContactDetails {
```
Nice reuse of ContactDetails

tahb commented on server/form-pages/apply/referrals-and-documents/accommodation-referral-details/crsDetails.test.ts:
```diff
@@ -0,0 +1,10 @@
+import { itShouldHaveNextValue, itShouldHavePreviousValue } from '../../../shared-examples'
+import CrsDetails from './crsDetails'
+
+describe('CrsDetails', () => {
```
Presumably we're not testing as much here because we trust that ContactDetails is being extended from. If that extension was removed then presumably our tests would continue to pass. I might consider checking that, can  we do something like this `expect(new CrsDetails({}, application) instanceOf ContactDetails).to eq(true)`?

bryangaledxw replied:
Hmm. The coverage checker would pick it up if his was doing more than just extending another class, but yeah, let's get that explicit check in

tahb reviewed (approved):
LGTM - a testing suggestion if it's possible

bryangaledxw committed:
Improve testing on classes that extend ContactDetails

We explicitly test that BackupContact and ProbationPractitioner extend ContactDetails so our tests will fail if this ever changes and we'd want more testing on these classes

selina-dxw commented:
Hey @bryangaledxw 

Couple of changes from me:

On the provide further details screen for the DTR and NOP can we change the date question to 'Date DTR / NOP was submitted'

And for the CRS referral we don't need the referrer details. The current referral form only asks if the CRS has been submitted, so we only need a yes / no answer. Sorry I missed that on Figma. 

Otherwise looks great.
